sql: Support privileges on virtual tables

RichardJCai opened the pull request:
Caching is added in this PR. `cacheutil.cache` is added to generalize the caching work done for role membership cache and userAuth cache. 

There is more testing work to be done on the cache + migration work to add an entry for `public` having `SELECT` on every virtual table.

Release note (sql change): Support privileges on virtual tables.
Previously users were unable to GRANT on virtual tables, ie
tables in crdb_internal / pg_catalog / information_schema.

Now users can GRANT/REVOKE `SELECT` privilege on virtual tables.
`SELECT` is needed to query a virtual table.

RichardJCai committed:
sql: Support privileges on virtual tables.

Release note (sql change): Support privileges on virtual tables.
Previously users were unable to GRANT on virtual tables, ie
tables in crdb_internal / pg_catalog / information_schema.

Now users can GRANT/REVOKE `SELECT` privilege on virtual tables.
`SELECT` is needed to query a virtual table.

Note that virtual tables privileges are NOT per database.
Doing `GRANT SELECT ON crdb_internal.tables TO foo` allows foo
to select on `crdb_internal.tables` across all databases.

Even if one does `GRANT SELECT ON dbname.crdb_internal.tables TO foo`
the database is ignored, there is a warning saying the database is
ignored.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @RichardJCai)

___
*[`pkg/sql/conn_executor_exec.go` line 1172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NWGmF2z_UokPy2FzR:-N7NWGmF2z_UokPy2FzS:b8sq04k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/conn_executor_exec.go#L1172)):*
> ```Go
> 		isSetOrShow := stmt.AST.StatementTag() == "SET" || stmt.AST.StatementTag() == "SHOW"
> 		if ex.sessionData().InjectRetryErrorsEnabled && !isSetOrShow &&
> 			planner.Txn().Sender().TxnStatus() == roachpb.PENDING && ex.user() != username.NodeUserName() {
> ```

hm i'm surprised this check for `node` is needed now
___
*[`pkg/sql/grant_revoke_system.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NWvy4C_CJpjMNz5oh:-N7NWvy4C_CJpjMNz5oi:bb0kqav) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/grant_revoke_system.go#L180)):*
> ```Go
> 
> 			if len(objectIDs) == 0 {
> 				return nil, errors.AssertionFailedf("no tables found")
> ```

does this need to error, or can we just show a NOTICE saying that nothing happened?
___
*[`pkg/sql/grant_revoke_system.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQ3oK28gzBBS73hVT:-N7NQ3oK28gzBBS73hVU:bx4awby) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/grant_revoke_system.go#L153)):*
> ```Go
> 		}
> 	}
> ```

we probably should have audit logs for grant and revoke here, to match the existing grant/revoke logs. could you file an issue for it?
___
*[`pkg/sql/planner.go` line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQGiX7DeroSdC69m2:-N7NQGiX7DeroSdC69m3:b-ioejlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/planner.go#L893)):*
> ```Go
> ) (descpb.DescriptorVersion, error) {
> 	_, desc, err := p.Descriptors().GetMutableTableByName(ctx, p.Txn(), name, flags)
> 	return desc.Version, err
> ```

nit: if err is non nil, then `desc` may be nil, and this could panic
___
*[`pkg/sql/user.go` line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQOyA5gv7vn4JaLGA:-N7NQOyA5gv7vn4JaLGB:bhsoss0) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/user.go#L470)):*
> ```Go
> 
> 	return p.writeSchemaChange(
> 		ctx, tableDesc, descpb.InvalidMutationID, "updating version for system privileges table",
> ```

nit: this shouldn't change
___
*[`pkg/sql/user.go` line 516 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQXeH8_4DhFgchDx3:-N7NQXeI0jeV_Mnsmwa-:b-394fr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/user.go#L516)):*
> ```Go
> // privileges table.
> func (p *planner) BumpPrivilegesTableVersion(ctx context.Context) error {
> 	privilegesTableName := tree.NewTableNameWithSchema("system", "public", "privileges")
> ```

nit: this could be a global var
___
*[`pkg/sql/user.go` line 523 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQccrAUn4CsLBgAYM:-N7NQccrAUn4CsLBgAYN:b-662yv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/user.go#L523)):*
> ```Go
> 
> 	return p.writeSchemaChange(
> 		ctx, tableDesc, descpb.InvalidMutationID, "updating version for role membership table",
> ```

nit: should say "system privileges table"
___
*[`pkg/sql/cacheutil/cache.go` line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NT7js8QSIvBo7eGwq:-N7NT7jt6IPTOwIqSb09:b-lx898r) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/cacheutil/cache.go#L50)):*
> ```Go
> // GetValue returns the value and if the key is found in the cache.
> // The cache lock must be held while calling this.
> func (c *Cache) GetValue(key interface{}) (interface{}, bool) {
> ```

nit: name it `GetValueLocked`
___
*[`pkg/sql/cacheutil/cache.go` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NUv4gCDun1XcFijL9:-N7NUv4gCDun1XcFijLA:b-7jzuon) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/cacheutil/cache.go#L120)):*
> ```Go
> // cached versions are newer, then false is returned to indicate that the
> // cached data should not be used.
> func (c *Cache) ClearCacheIfStale(
> ```

i think this needs a comment and rename to indicate it can only be called while the lock is held
___
*[`pkg/sql/cacheutil/cache.go` line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NVtlx6rnQOnTcSepQ:-N7NVtlx6rnQOnTcSepR:b-bzia71) ([raw file](https://github.com/cockroachdb/cockroach/blob/951f2ee1787b1248acb7045ebb609b170d716916/pkg/sql/cacheutil/cache.go#L41)):*
> ```Go
> // NewCache initializes a new Cache.
> func NewCache(
> 	account mon.BoundAccount, stopper *stop.Stopper, tableVersions []descpb.DescriptorVersion,
> ```

instead of a slice, can this param be an int for "number of backing system tables"
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NXuqFE4PHmAa4NYW3:-N7NXuqFE4PHmAa4NYW4:b-q0sya2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/catalog/tabledesc/validate.go#L515)):*
> ```Go
> 	if !desc.IsVirtualTable() {
> 		// Validate the privilege descriptor.
> 		if desc.Privileges == nil {
> ```

does the nil check need to happen before `IsVirtualTable`?
___
*[`pkg/sql/syntheticprivilege/synthesize_privileges.go` line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NU9NM6TlBeZAZfUp-:-N7NU9NM6TlBeZAZfUp0:bduq7ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/syntheticprivilege/synthesize_privileges.go#L38)):*
> ```Go
> 		cache.Lock()
> 		defer cache.Unlock()
> 		systemPrivilegesTableName := tree.NewTableNameWithSchema("system", "public", "privileges")
> ```

nit: could use a global var
___
*[`pkg/sql/syntheticprivilege/synthesize_privileges.go` line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NUXWrEQ3WWMJFJiLJ:-N7NUXWrEQ3WWMJFJiLK:bpfwa1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/syntheticprivilege/synthesize_privileges.go#L109)):*
> ```Go
> 		if systemTablePrivilegeObject.PrivilegeObjectType() == privilege.VirtualTable {
> 			if _, found := privileges.FindUser(username.PublicRoleName()); !found {
> 				privileges.Grant(username.PublicRoleName(), privilege.List{privilege.SELECT}, false)
> ```

nit: add inline comment for `false`
___
*[`pkg/sql/syntheticprivilege/vtable_privilege.go` line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYKn9CjVA6fTlL3bg:-N7NYKn9CjVA6fTlL3bh:b-el9jf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/syntheticprivilege/vtable_privilege.go#L69)):*
> ```Go
> // GetName implements the PrivilegeObject interface.
> func (p *VirtualTablePrivilege) GetName() string {
> 	// TODO(richardjcai): Make this get the actual table name.
> ```

how would we do this? just store the name alongside the ID?
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYTMg5xAQ9q5YNF4u:-N7NYTMg5xAQ9q5YNF4v:bxvanop) ([raw file](https://github.com/cockroachdb/cockroach/blob/262dabe03eb50801f89fd5c731449e725958a6fa/pkg/sql/logictest/testdata/logic_test/information_schema#L164)):*
> ```
> DROP INDEX information_schema.tables@i
> 
> statement error tables is a virtual object and cannot be modified
> ```

GRANT CREATE no vtables should still fail right?
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYcDv5zTuKbsd6C1f:-N7NYcDv5zTuKbsd6C1g:bz645bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
> ```
> 
> query B
> select has_table_privilege('testuser', 'crdb_internal.tables', 'select');
> ```

nit: add a query for `has_table_priv` before doing the REVOKE too
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYih52uwAwLwuMV3k:-N7NYih52uwAwLwuMV3l:bidvfi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L36)):*
> ```
> 
> query B
> select has_table_privilege('testuser', 'crdb_internal.tables', 'select');
> ```

at this point does system.privileges have no row, or a row with SELECT for the public role?
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N7NYwEO-I2f8ffqsexg:bl2yi95) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
> ```Go
> 				// tables due to system.privileges not having
> 				// a cache. We cannot perform the query without
> 				// the cache while the system range is out.
> ```

this should be added back now that there's a cache. also, i don't follow why this SHOW command was affected by vtable privs


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 1172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NWGmF2z_UokPy2FzR:-N7SLpvH0081ZexbR2PI:bpaa7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/conn_executor_exec.go#L1172)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm i'm surprised this check for `node` is needed now
</blockquote></details>

It originally was cause sometimes the privilege query would fail but I special cased it such that `Node` just short-circuits the checks so it should be okay now to remove.
___
*[`pkg/sql/grant_revoke_system.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NWvy4C_CJpjMNz5oh:-N7SQ5iQ-IcHqI7frsIE:b1yvllj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/grant_revoke_system.go#L180)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does this need to error, or can we just show a NOTICE saying that nothing happened?
</blockquote></details>

I don't think we can actually get in this state, even `GRANT SELECT ON crdb_internal.*` should always return a table. I don't think you can craft a glob that returns no tables given that all virtual schemas should have tables so I think we should error here.
___
*[`pkg/sql/grant_revoke_system.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQ3oK28gzBBS73hVT:-N7SMIJ99f2qLOMDJ9G_:b-c10ki) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/grant_revoke_system.go#L153)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we probably should have audit logs for grant and revoke here, to match the existing grant/revoke logs. could you file an issue for it?
</blockquote></details>

https://github.com/cockroachdb/cockroach/issues/84760

Done I'll update the issue details after.
___
*[`pkg/sql/planner.go` line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQGiX7DeroSdC69m2:-N7SMPak7mMEVqmizfrJ:b-h990k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/planner.go#L893)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: if err is non nil, then `desc` may be nil, and this could panic
</blockquote></details>

Good call. Fixed.
___
*[`pkg/sql/user.go` line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQOyA5gv7vn4JaLGA:-N7SQZ4P6v-byGmkjm2p:be9t6br) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/user.go#L470)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this shouldn't change
</blockquote></details>

Oops updated the wrong one
___
*[`pkg/sql/user.go` line 516 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NQXeH8_4DhFgchDx3:-N7SQwo25I5_VkPrNv37:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/user.go#L516)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this could be a global var
</blockquote></details>

Done
___
*[`pkg/sql/cacheutil/cache.go` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NUv4gCDun1XcFijL9:-N7SRqce0eiMzrfC0ojS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/cacheutil/cache.go#L120)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this needs a comment and rename to indicate it can only be called while the lock is held
</blockquote></details>

Done.
___
*[`pkg/sql/cacheutil/cache.go` line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NVtlx6rnQOnTcSepQ:-N7SR3ye6h92Zka3jpeW:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/951f2ee1787b1248acb7045ebb609b170d716916/pkg/sql/cacheutil/cache.go#L41)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of a slice, can this param be an int for "number of backing system tables"
</blockquote></details>

Done
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NXuqFE4PHmAa4NYW3:-N7SMq3Q1Jz6pBmai3-q:b8twv05) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/catalog/tabledesc/validate.go#L515)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does the nil check need to happen before `IsVirtualTable`?
</blockquote></details>

No it is now valid for virtual tables to not have a PrivilegeDescriptor
___
*[`pkg/sql/syntheticprivilege/synthesize_privileges.go` line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NUXWrEQ3WWMJFJiLJ:-N7SNL5Y-jvDkWREsM3W:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/74e4fcbbee918ed847ea93a6b42c4fae407cf127/pkg/sql/syntheticprivilege/synthesize_privileges.go#L109)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add inline comment for `false`
</blockquote></details>

Done
___
*[`pkg/sql/syntheticprivilege/vtable_privilege.go` line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYKn9CjVA6fTlL3bg:-N7SNTqIB1JSb-pI0Jp1:bsclqoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/syntheticprivilege/vtable_privilege.go#L69)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

how would we do this? just store the name alongside the ID?
</blockquote></details>

I've actually updated this to only use name instead of ID, currently IDs are not stable + names should never change for virtual tables.
___
*[`pkg/sql/logictest/testdata/logic_test/information_schema` line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYTMg5xAQ9q5YNF4u:-N7SNeqo9e7e0wbpExZC:b-66ho22) ([raw file](https://github.com/cockroachdb/cockroach/blob/262dabe03eb50801f89fd5c731449e725958a6fa/pkg/sql/logictest/testdata/logic_test/information_schema#L164)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

GRANT CREATE no vtables should still fail right?
</blockquote></details>

Yeah but I'm going to add this case to `virtual_table_privileges` 
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYcDv5zTuKbsd6C1f:-N7SO2jVBGUYCOkt-5Aj:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add a query for `has_table_priv` before doing the REVOKE too
</blockquote></details>

Done
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYih52uwAwLwuMV3k:-N7SO3OjADSQjmF0ow2k:b-r3jpic) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

at this point does system.privileges have no row, or a row with SELECT for the public role?
</blockquote></details>

It has a row - updated the test case
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N7SRvsZ3vXbyS94eoTf:bk6l289) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should be added back now that there's a cache. also, i don't follow why this SHOW command was affected by vtable privs
</blockquote></details>

The show command queries a vtable which we now need to query `system.privileges` to check.



_Code quote:_
```Go
				var isSuperuser string
				require.NoError(t, dbSQL.QueryRow(ctx, "SHOW is_superuser").Scan(&isSuperuser))
				require.Equal(t, "on", isSuperuser)
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
mostly small comments, looking good!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-:-N7bi2nT8LKs1Wn_ZB9-:b-gs7kud)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 186 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bcxGK2hMLiLedkYU6:-N7bcxGK2hMLiLedkYU7:b-x77u62) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/grant_revoke_system.go#L186)):*
> ```Go
> 
> 			for _, name := range tableNames {
> 				if !name.ExplicitCatalog {
> ```

nit: shouldn't this be `if name.ExplicitCatalog` so we show the notice if an explicit catalog was given?
___
*[`pkg/sql/cacheutil/cache.go` line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bdJwI0qpBrOnjVMQq:-N7bdJwI0qpBrOnjVMQr:bq5z43i) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/cacheutil/cache.go#L46)):*
> ```Go
> 	for i := 0; i < numSystemTables; i++ {
> 		tableVersions = append(tableVersions, 0)
> 	}
> ```

nit: `tableVersions := make([]descpb.DescriptorVersion, numSystemTables)` would allocate memory all at once, and will initialize to zeros.
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bh8oj-VYP5w2Qqonp:-N7bh8oj-VYP5w2Qqonq:bbdvacj) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
> ```
> 
> query TTTT
> SELECT * FROM system.privileges
> ```

is it worth adding an optimization so that if the row is just public having SELECT, then instead we delete the row instead? it could potentially make it easier to reason about the current state when debugging things
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bhxOm0rdMblmc4xEv:-N7bhxOm0rdMblmc4xEw:b-56j9ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L46)):*
> ```
> ----
> false
> ```

could you do a `SHOW TABLES` here and verify that it returns the right privilege error?
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 79 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bgqg20uBPUtGL7hfU:-N7bgqg33snFpsG1LSUV:b-tczjkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L79)):*
> ```
> CREATE DATABASE test2
> 
> statement ok
> ```

you can use `noticetrace` to make sure the warning is shown
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N7bf9anAuFZzz2NqVsF:b-cvubw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

The show command queries a vtable which we now need to query `system.privileges` to check.
</blockquote></details>

damn interesting.. tbh, that doesn't seem ideal since IMO a user should always be able to view their own session vars. and even more confusing since they will always have SET permission, but maybe won't have SHOW permission.

and also not ideal that SHOW now can result in a round trip. it would make debugging a loss of quorum incident harder.

i see that `sql/show_var.go` does not actually hit the vtable, but `delegate/show_var.go` does. maybe we should just do this diff so the delegate is only used for `show all`

```
diff --git a/pkg/sql/delegate/show_var.go b/pkg/sql/delegate/show_var.go
index e9e24d4700..a8fd954286 100644
--- a/pkg/sql/delegate/show_var.go
+++ b/pkg/sql/delegate/show_var.go
@@ -27,7 +23,6 @@ var ValidVars = make(map[string]struct{})

 // Show a session-local variable name.
 func (d *delegator) delegateShowVar(n *tree.ShowVar) (tree.Statement, error) {
-    origName := n.Name
     name := strings.ToLower(n.Name)

     if name == "locality" {
@@ -39,20 +34,4 @@ func (d *delegator) delegateShowVar(n *tree.ShowVar) (tree.Statement, error) {
             "SELECT variable, value FROM crdb_internal.session_variables WHERE hidden = FALSE",
         )
     }
-
-    if _, ok := ValidVars[name]; !ok {
-        // Custom options go to planNode.
-        if strings.Contains(name, ".") {
-            return nil, nil
-        }
-        return nil, pgerror.Newf(pgcode.UndefinedObject,
-            "unrecognized configuration parameter %q", origName)
-    }
-
-    varName := lexbase.EscapeSQLString(name)
-    nm := tree.Name(name)
-    return parse(fmt.Sprintf(
-        `SELECT value AS %[1]s FROM crdb_internal.session_variables WHERE variable = %[2]s`,
-        nm.String(), varName,
-    ))
 }
```

kind of an aside from all this, but this makes me realize that it can be non-obvious to users when a SHOW command delegates to a vtable, so in general it will be hard to see which privilege is for which SHOW command


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, and @rafiss)

___
*[`pkg/sql/grant_revoke_system.go` line 186 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bcxGK2hMLiLedkYU6:-N7bkq1J5gxT9VmRvqJv:bg5o82u) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/grant_revoke_system.go#L186)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: shouldn't this be `if name.ExplicitCatalog` so we show the notice if an explicit catalog was given?
</blockquote></details>

Oops yeah definitely, also not a nit :cold_face: 

Adding a test
___
*[`pkg/sql/cacheutil/cache.go` line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bdJwI0qpBrOnjVMQq:-N7bl38l6UjV04Z_B9If:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/cacheutil/cache.go#L46)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: `tableVersions := make([]descpb.DescriptorVersion, numSystemTables)` would allocate memory all at once, and will initialize to zeros.
</blockquote></details>

Done
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bh8oj-VYP5w2Qqonp:-N7blEb-8XozP4UTPQAU:b-n9f0q5) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is it worth adding an optimization so that if the row is just public having SELECT, then instead we delete the row instead? it could potentially make it easier to reason about the current state when debugging things
</blockquote></details>

Sure, I'm thinking we can also do the migration as well but if I'm keeping this for now, then this is a good idea.
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bhxOm0rdMblmc4xEv:-N7brKwN5E8eG6T0h2OH:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L46)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you do a `SHOW TABLES` here and verify that it returns the right privilege error?
</blockquote></details>

Done



_Code quote:_
```
select has_table_privilege('testuser', 'crdb_internal.tables', 'select')
```
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 79 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bgqg20uBPUtGL7hfU:-N7brZ3r2gR0etb30pXt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L79)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you can use `noticetrace` to make sure the warning is shown
</blockquote></details>

Done.
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N7br_E54QGqxu-35uvK:bfwood3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

damn interesting.. tbh, that doesn't seem ideal since IMO a user should always be able to view their own session vars. and even more confusing since they will always have SET permission, but maybe won't have SHOW permission.

and also not ideal that SHOW now can result in a round trip. it would make debugging a loss of quorum incident harder.

i see that `sql/show_var.go` does not actually hit the vtable, but `delegate/show_var.go` does. maybe we should just do this diff so the delegate is only used for `show all`

```
diff --git a/pkg/sql/delegate/show_var.go b/pkg/sql/delegate/show_var.go
index e9e24d4700..a8fd954286 100644
--- a/pkg/sql/delegate/show_var.go
+++ b/pkg/sql/delegate/show_var.go
@@ -27,7 +23,6 @@ var ValidVars = make(map[string]struct{})

 // Show a session-local variable name.
 func (d *delegator) delegateShowVar(n *tree.ShowVar) (tree.Statement, error) {
-    origName := n.Name
     name := strings.ToLower(n.Name)

     if name == "locality" {
@@ -39,20 +34,4 @@ func (d *delegator) delegateShowVar(n *tree.ShowVar) (tree.Statement, error) {
             "SELECT variable, value FROM crdb_internal.session_variables WHERE hidden = FALSE",
         )
     }
-
-    if _, ok := ValidVars[name]; !ok {
-        // Custom options go to planNode.
-        if strings.Contains(name, ".") {
-            return nil, nil
-        }
-        return nil, pgerror.Newf(pgcode.UndefinedObject,
-            "unrecognized configuration parameter %q", origName)
-    }
-
-    varName := lexbase.EscapeSQLString(name)
-    nm := tree.Name(name)
-    return parse(fmt.Sprintf(
-        `SELECT value AS %[1]s FROM crdb_internal.session_variables WHERE variable = %[2]s`,
-        nm.String(), varName,
-    ))
 }
```

kind of an aside from all this, but this makes me realize that it can be non-obvious to users when a SHOW command delegates to a vtable, so in general it will be hard to see which privilege is for which SHOW command
</blockquote></details>

I don't think the error is that unuitivitve, at least it clearly tells you which table.


<!-- Sent from Reviewable.io -->

postamar reviewed (changes requested):
I've done a first pass and also have a few comments and requests, nothing major.

> Reviewed 15 of 40 files at r1, 11 of 20 files at r4, 1 of 21 files at r5, 1 of 9 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-:-N7qDlWy3JCOjRXnO1Hj:bgoe986)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, @rafiss, and @RichardJCai)

___
*[`pkg/sql/catalog/privilege_object.go` line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q7HQ1DLw1p7xBQsjQ:-N7q7HQ1DLw1p7xBQsjR:bw1p9qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/catalog/privilege_object.go#L29)):*
> ```Go
> 	GetPrivilegeDescriptor(ctx context.Context, planner eval.Planner) (*catpb.PrivilegeDescriptor, error)
> 	// GetObjectTypeName returns user facing name of the object type.
> 	GetObjectTypeName() string
> ```

I'm curious, does this method even need to exist? It seems like its call sites could be replaced by `GetObjectType()` cast to `string`. I'm probably missing something. It would be nice to get rid of this method, though.
___
*[`pkg/sql/catalog/tabledesc/table_desc.go` line 624 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7qDGvQ5SAA47DjA-v3:-N7qDGvQ5SAA47DjA-v4:behib9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d985322bd79a7d53cc0ef2d1ec48d3b729eee106/pkg/sql/catalog/tabledesc/table_desc.go#L624)):*
> ```Go
> 		// do not store a database id, so the descriptors are not
> 		// "per database" even though regular tables are per database.
> 		vsName, found := planner.GetVirtualSchemaNameByID(desc.GetParentSchemaID())
> ```

You don't need this method in `eval.Planner` at all, you cat get the schema name by looking up `catconstants.StaticSchemaIDMap` or calling `schemadesc.GetVirtualSchemaByID`.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 522 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q7vj36EP0brghns-i:-N7q7vj36EP0brghns-j:b-5g0eys) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/catalog/tabledesc/validate.go#L522)):*
> ```Go
> 			} else {
> 				vea.Report(catprivilege.Validate(*desc.Privileges, desc, privilege.Table))
> 			}
> ```

Could this if-block be replaced with 
```
vea.Report(catprivilege.Validate(*desc.Privileges, desc, desc.GetObjectType()))
```
___
*[`pkg/sql/sem/eval/deps.go` line 330 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q8pVS5DscJkbElsVd:-N7q8pVS5DscJkbElsVe:b4zupvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/sem/eval/deps.go#L330)):*
> ```Go
> 
> 	// IsActive returns if the version specified by key is active.
> 	IsActive(ctx context.Context, key clusterversion.Key) bool
> ```

Do you really need to add this method? This interface already gives you access to the eval context.
___
*[`pkg/sql/sem/eval/deps.go` line 339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7qAgDy7CR6sXNLTM1s:-N7qAgDy7CR6sXNLTM1t:b-psviwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/34694b200b1501966b23531e95f145aee7740e73/pkg/sql/sem/eval/deps.go#L339)):*
> ```Go
> 	// GetDescriptorVersionByTableName returns the descriptor version of a
> 	// given table.
> 	GetDescriptorVersionByTableName(ctx context.Context, name tree.ObjectName, flags tree.ObjectLookupFlags) (descpb.DescriptorVersion, error)
> ```

I'm not sure that these are the kinds of methods we want in the `eval.Planner` interface. We certainly shouldn't be too casual about adding stuff there, it comes at a cost.

To me it seems you'd be better off exposing a method here that takes a `systemTablePrivilegeObject` string and returns a `(*catpb.PrivilegeDescriptor, error)`.


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: add caching for synthetic privileges

Also minor moving around of code

Release note: None

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, and @rafiss)

___
*[`pkg/sql/catalog/privilege_object.go` line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q7HQ1DLw1p7xBQsjQ:-N7wH31z-oxBwJla8ZWc:b-y70io6) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/catalog/privilege_object.go#L29)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm curious, does this method even need to exist? It seems like its call sites could be replaced by `GetObjectType()` cast to `string`. I'm probably missing something. It would be nice to get rid of this method, though.
</blockquote></details>

Yeah this is unnecessary. Removed.
___
*[`pkg/sql/catalog/tabledesc/table_desc.go` line 624 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7qDGvQ5SAA47DjA-v3:-N7wJCdn9JVtGFScFsa0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d985322bd79a7d53cc0ef2d1ec48d3b729eee106/pkg/sql/catalog/tabledesc/table_desc.go#L624)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You don't need this method in `eval.Planner` at all, you cat get the schema name by looking up `catconstants.StaticSchemaIDMap` or calling `schemadesc.GetVirtualSchemaByID`.
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 522 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q7vj36EP0brghns-i:-N7wWfM12LiUju7q59f1:b-c7dfnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/catalog/tabledesc/validate.go#L522)):*
> vea.Report(catprivilege.Validate(*desc.Privileges, desc, desc.GetObjectType()))

Yep, good call
___
*[`pkg/sql/sem/eval/deps.go` line 330 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7q8pVS5DscJkbElsVd:-N7wXU3-0G5nYUW43Mfn:bnzopt) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a0627a72231c02fcf357716932beee3cffeee6/pkg/sql/sem/eval/deps.go#L330)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Do you really need to add this method? This interface already gives you access to the eval context.
</blockquote></details>

I thought this interface is used from the EvalContext and not the other way around.

I might be missing something but I don't see how to check the version with just a `planner eval.Planner`.

Although I suppose since I've moved `SynthesizePrivilege...` as a method of planner, I could make it return nil and that can indicate which version we're on? Is that preferable?
___
*[`pkg/sql/sem/eval/deps.go` line 339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7qAgDy7CR6sXNLTM1s:-N7wYmZ85Pcl4WrX5MbT:b-92ldt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/34694b200b1501966b23531e95f145aee7740e73/pkg/sql/sem/eval/deps.go#L339)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm not sure that these are the kinds of methods we want in the `eval.Planner` interface. We certainly shouldn't be too casual about adding stuff there, it comes at a cost.

To me it seems you'd be better off exposing a method here that takes a `systemTablePrivilegeObject` string and returns a `(*catpb.PrivilegeDescriptor, error)`.
</blockquote></details>

Fair enough, I'll add the synthesizePrivilegeDescriptor method onto the planner interface here


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, @rafiss, and @RichardJCai)

___
*[`pkg/sql/information_schema.go` line 1372 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kTMw0wVPL9r5xvy6:-N8-kTMw0wVPL9r5xvy7:bvxr5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/information_schema.go#L1372)):*
> ```Go
> 			// TODO(knz): This should filter for the current user, see
> 			// https://github.com/cockroachdb/cockroach/issues/35572
> 			//var tableType privilege.ObjectType
> ```

i think this was commented accidentally
___
*[`pkg/sql/planner.go` line 889 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kcsy7Z9vpCDG5BIi:-N8-kcsy7Z9vpCDG5BIj:bn8dqv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L889)):*
> ```Go
> 
> // SynthesizePrivilegeDescriptorFromSystemPrivilegesTable is part of the Planner interface.
> func (p *planner) SynthesizePrivilegeDescriptorFromSystemPrivilegesTable(
> ```

nit: is "FromSystemPrivilegesTable" really needed as part of the name?

also nit: maybe this would fit better in `grant_revoke_system.go` so that this file doesn't get too big.
___
*[`pkg/sql/planner.go` line 901 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-n3UtDVOt_cw8kzH1:-N8-n3Uu5VI6O8i0f_kJ:bwn1a5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L901)):*
> ```Go
> 		cache.Lock()
> 		defer cache.Unlock()
> 		systemPrivilegesTableName := tree.NewTableNameWithSchema("system", "public", "privileges")
> ```

nit: this can use the global var
___
*[`pkg/sql/planner.go` line 935 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-nxFa6RxwNzoPSd-h:-N8-nxFbA6d7O-lTN1CK:b-87yhlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L935)):*
> ```Go
> 		}
> 		defer func() {
> 			retErr = errors.CombineErrors(retErr, it.Close())
> ```

i think retErr is wrong here. right now it's referring to the return value of SynthesizePrivilegeDescriptorFromSystemPrivilegesTable

i think it should refer to the return value of `func(loadCtx context.Context) (interface{}, error)`
___
*[`pkg/sql/planner.go` line 993 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kr1h0PkoUcILBJ_9:-N8-kr1i5vmF97LHOMyA:b9ibbdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L993)):*
> ```Go
> 		// grant. If there is an empty row for Public, then public
> 		// does not have grant.
> 		// TODO(richardjcai): Do a migration for this instead.
> ```

are we sure we want a migration for it?
___
*[`pkg/sql/cacheutil/cache.go` line 43 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-l2Qu5Opy-z1CjMic:-N8-l2Qu5Opy-z1CjMid:bonakjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/cacheutil/cache.go#L43)):*
> ```Go
> // We use it to initialize the tableVersions slice to 0 for each table.
> func NewCache(account mon.BoundAccount, stopper *stop.Stopper, numSystemTables int) *Cache {
> 	var tableVersions []descpb.DescriptorVersion
> ```

copying over older comment:

nit: tableVersions := make([]descpb.DescriptorVersion, numSystemTables) would allocate memory all at once, and will initialize to zeros.
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bh8oj-VYP5w2Qqonp:-N8-m5Me37xktwAKRYiM:b-giz68v) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Sure, I'm thinking we can also do the migration as well but if I'm keeping this for now, then this is a good idea.
</blockquote></details>

i thought we were saying we'd want to avoid the migration. otherwise, each new vtable will also require a migration.
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N8-mMVo8_w_F-5-7AK4:b-k4evev) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I don't think the error is that unuitivitve, at least it clearly tells you which table.
</blockquote></details>

OK. i still think this PR should apply the diff above so that `SHOW  var` does not require round trips in any cases. (and remove this extra SHOW that was added to the test)
___
*[`pkg/sql/privilegeobject/privilege_object.go` line 23 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-ld4pDLUeo8mw0bUY:-N8-ld4q8Uaa9bSxC32h:b-jv0908) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/privilegeobject/privilege_object.go#L23)):*
> ```Go
> // PrivilegeObject represents an object that can have privileges. The privileges
> // can either live on the descriptor or in the system.privileges table.
> type PrivilegeObject interface {
> ```

why did this need to move to a new package? just wondering since `privilegeObject := privilegeobject.PrivilegeObject` in other parts of the code seems a bit verbose


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: shuffle PrivilegeObject and syntheticPrivilegeObject packages

Release note: None

RichardJCai committed:
sql: update delegate/show_var.go such that only SHOW ALL queries crdb_internal.

sql/show_var.go does not query crdb_internal whereas delegate does.
This update makes us not make unncessary roundtrips.

Release note: None

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, and @rafiss)

___
*[`pkg/sql/information_schema.go` line 1372 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kTMw0wVPL9r5xvy6:-N8-rXGo9wL91Dl_7bSL:bkp68c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/information_schema.go#L1372)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this was commented accidentally
</blockquote></details>

Removed commented code
___
*[`pkg/sql/planner.go` line 889 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kcsy7Z9vpCDG5BIi:-N8-rw8T-j2lftC0guOp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L889)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: is "FromSystemPrivilegesTable" really needed as part of the name?

also nit: maybe this would fit better in `grant_revoke_system.go` so that this file doesn't get too big.
</blockquote></details>

Done.
___
*[`pkg/sql/planner.go` line 935 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-nxFa6RxwNzoPSd-h:-N8-sIui93U0ykQZe41f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L935)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think retErr is wrong here. right now it's referring to the return value of SynthesizePrivilegeDescriptorFromSystemPrivilegesTable

i think it should refer to the return value of `func(loadCtx context.Context) (interface{}, error)`
</blockquote></details>

Done.
___
*[`pkg/sql/planner.go` line 993 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kr1h0PkoUcILBJ_9:-N8-sLmU0Ldyu9H3-ol0:b-1mo5bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/planner.go#L993)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

are we sure we want a migration for it?
</blockquote></details>

I think I'm happy with just empty row = defualt case.
___
*[`pkg/sql/cacheutil/cache.go` line 43 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-l2Qu5Opy-z1CjMic:-N8-sV8N5n0MyqKtsk9b:by7bqsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/cacheutil/cache.go#L43)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

copying over older comment:

nit: tableVersions := make([]descpb.DescriptorVersion, numSystemTables) would allocate memory all at once, and will initialize to zeros.
</blockquote></details>

Oops I think I reverted some things accidentally when rebasing
___
*[`pkg/sql/logictest/testdata/logic_test/virtual_table_privileges` line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7bh8oj-VYP5w2Qqonp:-N8-so-oDx7pPniaL4s6:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/00ffd6c608614fcef112051d2c7c3ded3951a442/pkg/sql/logictest/testdata/logic_test/virtual_table_privileges#L28)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i thought we were saying we'd want to avoid the migration. otherwise, each new vtable will also require a migration.
</blockquote></details>

Ack.
___
*[`pkg/sql/user_test.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N7NYwEO-I2f8ffqsexf:-N80231W4SgZ1r5yNdxD:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c83acf4c9a2495506507de95040f488b1ba7e72/pkg/sql/user_test.go#L173)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

OK. i still think this PR should apply the diff above so that `SHOW  var` does not require round trips in any cases. (and remove this extra SHOW that was added to the test)
</blockquote></details>

Done
___
*[`pkg/sql/privilegeobject/privilege_object.go` line 23 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-ld4pDLUeo8mw0bUY:-N8-xlerDbHsY1rg5Z-R:bjeokih) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/privilegeobject/privilege_object.go#L23)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why did this need to move to a new package? just wondering since `privilegeObject := privilegeobject.PrivilegeObject` in other parts of the code seems a bit verbose
</blockquote></details>

I thought this would be more appropriate, but yeah I think this is verbose and the duplication of naming is bad here. I've moved it back to `catalog` But I did move `SyntheticPrivilegeObject` into `syntheticprivilege` package and named it `Object`


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 228 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N80VnQuB1rElMEVzs5S:-N80VnQuB1rElMEVzs5T:bhic5h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2121f666f0ff1e2bac53433d5d7c25fcddc4c4a4/pkg/sql/grant_revoke_system.go#L228)):*
> ```Go
> 	privilegeObjectPath string,
> 	privilegeObjectType privilege.ObjectType,
> ) (privileges *catpb.PrivilegeDescriptor, retErr error) {
> ```

i think this `retErr` doesn't actually need to be named.
___
*[`pkg/sql/grant_revoke_system.go` line 263 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N80W2aO4HuUu_ItgeBz:-N80W2aO4HuUu_ItgeC-:b1fdg6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2121f666f0ff1e2bac53433d5d7c25fcddc4c4a4/pkg/sql/grant_revoke_system.go#L263)):*
> ```Go
> 			privilegeObjectPath)
> 
> 		it, err := p.QueryIteratorEx(ctx, `get-system-privileges`,
> ```

err is a named return value, but is also shadowed here. i think it would be more clear to use `retErr` for the return value
___
*[`pkg/sql/information_schema.go` line 1372 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8-kTMw0wVPL9r5xvy6:-N80UjsW0astzPHXdUxX:b-6x49zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52861c674b00d83f9c9cf22d8b91ddae2bf1abf/pkg/sql/information_schema.go#L1372)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Removed commented code
</blockquote></details>

the code should not be commented or removed. that code is meaningful. your PR first commented and then removed it


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-:-N80Z1hv4h_NKky0d8UU:b-7m74ox)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/sem/tree/show.go:
```diff
@@ -599,15 +599,6 @@ func (node *ShowSyntax) Format(ctx *FmtCtx) {
 }
 
-// ShowTransactionStatus represents a SHOW TRANSACTION STATUS statement.
-type ShowTransactionStatus struct {
```
we still need `ShowTransactionStatus` as a special case, since it's an observer statement (i.e. doesn't use the planner at all)

rafiss commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -1745,8 +1745,6 @@ func (ex *connExecutor) runObserverStatement(
 	ctx context.Context, ast tree.Statement, res RestrictedCommandResult,
 ) error {
 	switch sqlStmt := ast.(type) {
-	case *tree.ShowTransactionStatus:
```
this is the important part to keep.

rafiss commented on pkg/sql/delegate/delegate.go:
```diff
@@ -142,9 +142,6 @@ func TryDelegate(
 	case *tree.ShowZoneConfig:
 		return d.delegateShowZoneConfig(t)
 
-	case *tree.ShowTransactionStatus:
```
this part should not be needed at all, so it can remain deleted

adityamaru commented on pkg/ccl/backupccl/restore_old_versions_test.go:
```diff
@@ -246,6 +246,7 @@ ORDER BY object_type, object_name`, [][]string{
 	})
 
 	t.Run("zoneconfig_privilege_restore", func(t *testing.T) {
+		skip.WithIssue(t, 84359)
```
drive by: 84359 has been resolved so would be good to unskip this if possible

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83604)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @postamar, @rafiss, and @RichardJCai)

___
*[`pkg/sql/delegate/show_var.go` line 47 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/83604#-N8aoEsNAY6hh6j5HBV6:-N8aoEsNAY6hh6j5HBV7:b-993d7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/af69feb4d7d2833a2255aa7d79f4c9c725b745ae/pkg/sql/delegate/show_var.go#L47)):*
> ```Go
> 	if name == "database" {
> 		return parse(
> 			"SELECT value FROM crdb_internal.session_variables WHERE variable = 'database'")
> ```

i think if you change this to `SELECT value AS database FROM crdb_internal.session_variables WHERE variable = 'database'` then the tests will pass.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
Thanks for reviewing!

bors r=rafiss

adityamaru commented:
No need to block this bors but would be nice to unskip https://github.com/cockroachdb/cockroach/pull/83604/files#r935981990 or understand why its skipped now that the linked issue has been fixed, thanks!

RichardJCai commented:
> No need to block this bors but would be nice to unskip https://github.com/cockroachdb/cockroach/pull/83604/files#r935981990 or understand why its skipped now that the linked issue has been fixed, thanks!

Sorry yeah I need to unskip that, will do so once this is merged.
